### PR TITLE
Update metrics

### PR DESCRIPTION
**Pull request by alisonmyers**
Added topic summary metrics, and the ability to summarize by module. Reorganized functions into util where appropriate and added some more error handling for checking the .env file.

### CODE REVIEW

File: writeSummaryByModuleToCSV.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  modules.forEach(module => {
+    const summary = moduleSummary(module)
+    const moduleDetails = {
+      module_id: module.id,
+      module_name: stripHTML(escapeComment(module.name)),
+      module_created_at: module.unlock_at || module.created_at,
+      number_of_posts: summary.numberOfPosts,
+      median_word_count: summary.medianWordCount,
+      average_time_to_post_hours: summary.averageTimeDiff,
+      first_reply_timestamp: summary.firstReplyTimestamp,
```
Comment (craigdsthompson) [inline writeSummaryByModuleToCSV.js:42]: `first_reply_timestamp` format is different from `module_created_at`

**Review by craigdsthompson — COMMENTED** [review]
Lots of suggestions for clarification and ways to tidy, feel free to push back on (or ignore) any suggestions you disagree with!

File: writeToCSV.js [hunk 2]
```diff
@@ -1,69 +1,62 @@
...
+  const headers =[
+    'topic_id', 
+    'topic_title', 
+    'topic_message', 
     'topic_author_id',
-    'topic_author_name',
-    'topic_timestamp',
+    'topic_author_name', 
+    'topic_created_at', 
+    'topic_posted_at',
```
[the remark this answers is not in the document]
Comment (alisonmyers) [reply writeToCSV.js:18]: Yes. I believe course copied discussions might not have posted_at information or topic author information 

  // The datetime the topic was posted. If it is null it hasn't been posted yet.
  // (see delayed_post_at)
  "posted_at": "2037-07-21T13:29:31Z",

File: util.js [hunk 3]
```diff
@@ -1,9 +1,100 @@
...
+const appendRow = (pathToFile, row) => fs.appendFileSync(pathToFile, row.join(',') + '\r\n')
+
+// Word count function
+const getWordCount = (str) => {
+  const cleanStr = stripHTML(escapeComment(str))
+  //const cleanStr = str.replace(/<\/?[^>]+(>|$)/g, "") // Remove HTML tags
+  return cleanStr.trim().split(/\s+/).length
+}
+
+// Median word count
```
[the remark this answers is not in the document]
Comment (alisonmyers) [reply util.js:22]: will address in new PR - uses an NLP library.

File: util.js [hunk 4]
```diff
@@ -1,9 +1,100 @@
...
+      averageTimeToPostFromFirst: 0,
+      averagePostsPerAuthor: 0
+    }
+  }
+
+
+  // Word counts
+  const wordCounts = posts.map(post => getWordCount(post.postMessage))
+
+  const medianWordCount = Math.round(median(wordCounts) * 10) / 10
```
[the remark this answers is not in the document]
Comment (alisonmyers) [reply util.js:49]: will address in new PR - see previous comment.

File: writeToCSV.js [hunk 5]
```diff
@@ -1,69 +1,62 @@
...
     'post_author_id',
-    'post_author_name',
-    'post_id',
-    'post_parent_id',
+    'post_author_name', 
+    'post_id',
```
[the remark this answers is not in the document]
Comment (alisonmyers) [reply writeToCSV.js:21]: Yes - I'm not sure why but this is behaviour possible in Canvas.

File: writeToCSV.js [hunk 6]
```diff
@@ -1,69 +1,62 @@
...
 
-  const header = [
-    'topic_id',
-    'topic_title',
-    'topic_message',
+  const headers =[
+    'topic_id', 
+    'topic_title', 
+    'topic_message', 
     'topic_author_id',
```
[the remark this answers is not in the document]
Comment (alisonmyers) [reply writeToCSV.js:15]: Yes - there are cases without author id or name, I think this is from course copied material

File: util.js [hunk 7]
```diff
@@ -1,9 +1,100 @@
...
+  const averagePostsPerAuthor = Math.round((Object.values(postCountsByAuthor).reduce((acc, curr) => acc + curr, 0) / Object.keys(postCountsByAuthor).length) * 10) / 10
+
+  
+  const firstReplyTimestamp = new Date(Math.min(...posts.map(post => new Date(post.postTimestamp))))
+  const timeDiffsFromFirst = posts
+    .map(post => {
+      const postTimestamp = new Date(post.postTimestamp)
+      return postTimestamp > firstReplyTimestamp 
+        ? (postTimestamp - firstReplyTimestamp) / (1000 * 60 * 60) // Convert from milliseconds to hours
+        : null
```
[the remark this answers is not in the document]
Comment (alisonmyers) [reply util.js:73]: there are cases with no postTimestamp

File: index.js [hunk 8]
```diff
@@ -33,48 +45,106 @@ const getNestedReplies = (replyObj, participants, topicId) => {
   }, ...replies]
 }
 
+const getDiscussionsAndTopics = async (courseId, topicIds) => {
+  const discussionsAndTopics = await Promise.all(
+    topicIds.map(async topicId => {
+      const discussion = await capi.getFullDiscussion(courseId, topicId)
```
[the remark this answers is not in the document]
Comment (alisonmyers) [reply index.js:58]: restructured to call synchronously and then await results before returning

File: index.js [hunk 9]
```diff
@@ -33,48 +45,106 @@ const getNestedReplies = (replyObj, participants, topicId) => {
...
-const courseIds = process.env.COURSE_IDS.split(',').map(id => id.trim());
+const courseIds = process.env.COURSE_IDS.split(',').map(id => id.trim())
+const returnSummaryByModule = process.env.INCLUDE_MODULE_SUMMARY ? process.env.INCLUDE_MODULE_SUMMARY === 'true' : false
 
 Promise.all(
-  courseIds.map(courseId =>
-    getDiscussions(courseId)
-      .then(discussions => writeToCSV(courseId, discussions))
-  )
+  courseIds.map(courseId => {
```
[the remark this answers is not in the document]
Comment (alisonmyers) [reply index.js:138]: Will create a new issue

File: index.js [hunk 10]
```diff
@@ -33,48 +45,106 @@ const getNestedReplies = (replyObj, participants, topicId) => {
...
+
+const getPublishedModuleDiscussions = async courseId => {
+  
+  const modules = await capi.getModules(courseId)
+
+  const modulesWithDiscussionItems = await Promise.all(modules.map(async module => {
+    const items = await capi.getModuleItems(courseId, module.id)
+    const discussionItems = items.filter(item => item.type === "Discussion" && item.published===true).flat()
+
+    const discussionsAndTopics = await getDiscussionsAndTopics(courseId, discussionItems.map(item => item.content_id))
```
[the remark this answers is not in the document]
Comment (alisonmyers) [reply index.js:113]: #32

File: util.js [hunk 11]
```diff
@@ -1,9 +1,100 @@
...
+  }
+
+
+  // Word counts
+  const wordCounts = posts.map(post => getWordCount(post.postMessage))
+
+  const medianWordCount = Math.round(median(wordCounts) * 10) / 10
+
+  // Average time in hours from topicCreatedAt to postTimestamp
+  const timeDiffs = posts.map(post => {
```
[the remark this answers is not in the document]
Comment (alisonmyers) [reply util.js:65]: will resolve in new PR where adding tests and refactoring some of the util functions.

**Comment by alisonmyers** [comment]
Thanks @craigdsthompson! All comments will be or have been addressed:  for a few of the bigger ones I created new issues. I have another PR in the works that uses `natural` from npm for some stats and I'm adding util function tests there and I have updated some naming styles for better readability. 

All timestamp outputs are now created when the file is created and use Vancouver timezone.

**Review by craigdsthompson — APPROVED** [review]
Looks good!

**Pull request merged**